cpp/internal/CMakeLists.txt

```diff
@@ -0,0 +1,21 @@
+# =============================================================================
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License. You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License
+# is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+# or implied. See the License for the specific language governing permissions and limitations under
+# the License.
+# =============================================================================
+
+if(BUILD_TESTS OR BUILD_BENCH)
+  add_library(raft_internal INTERFACE)
```

Comment:
I'm particularly not sure and welcome any suggestions regarding the naming of the component. Currently:
  1. The name is `raft_internal`, because I wanted to call it `internal`, but it may be a reserved word in cmake.
  2. All headers are placed into double-nested `internal/internal`. The first `internal` is for cmake to isolate it from the other components, the second is for includes, so that all includes from this component are easily identifiable, e.g.:
  ```C++
  #include <internal/neighbors/naive_knn.cuh>
  ```

Reply:
Thinking through this a little bit- what do you think about keeping the outer directory `cpp/internal` as-is but nesting everything inside `cpp/internal/raft` instead of `cpp/internal/internal`. This would effectively allows us to include the `cpp/internal` directory in the build for the bench and test source files whilst making the include paths themselves look the same as the other raft includes. So for example, this would still allow us to include the `naive_knn.cuh` file as `#include <raft/neighbors/naive_knn.cuh>` but the file would be coming from `cpp/internal` instead. This assumes we wouldn't have name collisions there, which I highly doubt we would. 

Another option could be to do something like `cpp/internal/raft/internal` but I kind of prefer the option above more. What do you think?

Reply:
Actually, it turns out we have name clashes already! In branch-23.02 we have `include/raft/matrix/select_k.cuh` and its corresponding test `test/raft/matrix/select_k.cuh`; the latter is included using quotes syntax, so it's not a problem so far. In this PR I moved it to the `internal` component and changed the include syntax, so it would be a problem if we change the folder name to raft. So I see two options:
  1. rename `test/raft/matrix/select_k.cuh` to something like `select_k_helpers.cuh` or `select_k_utils.cuh`
  2. rename the root folder `internal` to something different from `raft`, yet not potentially clashing with third-party libraries.
 
In the new commit I opted for the option (2) with `raft_internal` as an experiment. Though I don't have a strong preference in this question. I only think that the trailing `internal` in the folder structure is not the best option, because it feels somewhat harder to maintain.

Reply:
I think I can live w/ `raft_internal` in the meantime. We can always revisit if we find this is presenting maintenance challenges.